src/Memory/MemHeap.cpp

```diff
@@ -16,20 +16,29 @@ namespace caffeine {
  ***************************************************/
 
 Allocation::Allocation(const OpRef& address, const OpRef& size,
-                       const OpRef& data, AllocationKind kind)
-    : address_(address), size_(size), data_(data), kind_(kind) {
+                       const OpRef& data, AllocationKind kind,
+                       AllocationPermissions permissions)
+    : address_(address), size_(size), data_(data), kind_(kind),
+      perms_(permissions) {
   CAFFEINE_ASSERT(address->type().is_int());
   CAFFEINE_ASSERT(size->type().is_int());
   CAFFEINE_ASSERT(address->type().bitwidth() == size->type().bitwidth());
 }
 Allocation::Allocation(const OpRef& address, const ConstantInt& size,
-                       const OpRef& data, AllocationKind kind)
-    : Allocation(address, make_ref<Operation>(size), data, kind) {}
+                       const OpRef& data, AllocationKind kind,
+                       AllocationPermissions permissions)
+    : Allocation(address, make_ref<Operation>(size), data, kind, permissions) {}
 
 void Allocation::overwrite(const OpRef& newdata) {
+  CAFFEINE_ASSERT(perms_ == AllocationPermissions::Write ||
+                      perms_ == AllocationPermissions::ReadWrite,
```

Comment:
Using the bitfield idea:
```suggestion
  CAFFEINE_ASSERT(perms_ & AllocationPermissions::Write,
```